src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/locations/LocationService.kt

```diff
@@ -49,7 +54,152 @@ class LocationService(
+    val location = agencyInternalLocationRepository.findByIdOrNull(locationId)
+      ?: throw NotFoundException("Location with id=$locationId does not exist")
+
+    val internalLocationType =
+      internalLocationTypeRepository.findByIdOrNull(InternalLocationType.pk(locationDto.locationType))
+        ?: throw BadDataException("Location type with id=${locationDto.locationType} does not exist")
+
+    val housingUnitType = locationDto.unitType?.let {
+      housingUnitTypeRepository.findByIdOrNull(HousingUnitType.pk(it))
+        ?: throw BadDataException("Housing unit type with id=${locationDto.unitType} does not exist")
+    }
+
+    val parent = locationDto.parentLocationId?.let {
+      agencyInternalLocationRepository.findByIdOrNull(it)
+        ?: throw BadDataException("Parent location with id=$it does not exist")
+    }
+
+    location.apply {
+      description =
+        location.parentLocation?.let { "${location.parentLocation!!.description}-${locationDto.locationCode}" }
+          ?: locationDto.locationCode
+      locationType = internalLocationType
+      description = locationDto.description
+      userDescription = locationDto.userDescription
+      locationCode = locationDto.locationCode
+      parentLocation = parent
+      listSequence = locationDto.listSequence
+      comment = locationDto.comment
+      unitType = housingUnitType
+    }.also {
+      telemetryClient.trackEvent(
+        "location-updated",
+        mapOf(
+          "locationId" to it.locationId.toString(),
+          "description" to it.description,
+        ),
+        null,
+      )
+    }
+  }
+
+  @Audit
+  fun deactivateLocation(locationId: Long, deactivateRequest: DeactivateRequest) {
+    val location = agencyInternalLocationRepository.findByIdOrNull(locationId)
+      ?: throw NotFoundException("Location with id=$locationId does not exist")
+
+    if (location.deactivateDate != null) {
+      if (location.reactivateDate == null || location.reactivateDate!!.isBefore(location.deactivateDate)) {
+        throw BadDataException("Location with id=$locationId is already deactivated")
+      }
+    }
+
+    location.deactivateDate = LocalDate.now()
+    location.deactivateReason = deactivateRequest.reasonCode?.let {
+      livingUnitReasonRepository.findByIdOrNull(LivingUnitReason.pk(it))
+        ?: throw BadDataException("Deactivate Reason code=$it does not exist")
+    }
+
+    telemetryClient.trackEvent(
+      "location-deactivated",
+      mapOf(
+        "locationId" to location.locationId.toString(),
+        "description" to location.description,
+      ),
+      null,
+    )
+  }
+
+  @Audit
+  fun reactivateLocation(locationId: Long) {
+    val location = agencyInternalLocationRepository.findByIdOrNull(locationId)
+      ?: throw NotFoundException("Location with id=$locationId does not exist")
+
+    if (location.deactivateDate == null ||
+      (location.reactivateDate != null && !location.reactivateDate!!.isBefore(location.deactivateDate))
```

Comment:
strange `!!` here - can never be null as previous condition prevents it?

Reply:
Well good Q ... its a shame but otherwise intellij gives an error "Smart cast to 'LocalDate' is impossible, because 'location.reactivateDate' is a mutable property that could have been changed by this time" i.e. its a var